File: Algorithms/StackAndQueue/LongestValidParentheses.py
"""
Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

For "(()", the longest valid parentheses substring is "()", which has length = 2.

Another example is ")()())", where the longest valid parentheses substring is "()()", which has length = 4.
"""

import unittest


class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        longest, last = 0, -1  # last: the last position of )
        stack = []             # the positions of non-matching (
        for i, c in enumerate(s):
            if c == '(':
                stack.append(i)
            else:
                if not stack:
                    last = i
                else:
                    stack.pop()
                    if not stack:
                        longest = max(longest, i - last)
                    else:
                        longest = max(longest, i - stack[-1])
        return longest

    # Dynamic Programming, One Pass
    # Time O(n), Space O(n)

    # Two-way scan
    # Time O(n), Space O(1)


class SolutionTest(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test1(self):
        string = ")()())"
        self.assertEqual(self.s.longestValidParentheses(string), 4)

    def test2(self):
        string = ")()())(((((()()()()("
        self.assertEqual(self.s.longestValidParentheses(string), 8)

    def test3(self):
        string = "()(())"
        self.assertEqual(self.s.longestValidParentheses(string), 6)


if __name__ == '__main__':
    unittest.main()

File: Algorithms/String/LongestValidParentheses.cpp
/**
 * Given a string containing just the characters '(' and ')', find the length of the longest valid (well-formed) parentheses substring.

Example 1:

Input: "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()"

Example 2:

Input: ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()"


 */

#include <string>
#include <iostream>
#include <stack>
using namespace std;


class Solution {
public:
    int longestValidParentheses(string s) {
        int longest = 0;
        stack<int> cstack;
        cstack.push(-1);
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                cstack.push(i);
            } else {
                cstack.pop();
                if (!cstack.empty()) {
                    int l = cstack.top();
                    longest = max(i - l, longest);
                } else {
                    cstack.push(i);
                }
            }
        }
        return longest;
    }
};


int main() {
    //string s = "()(()";
    string s = "()(())";
    Solution sol;
    int ret = sol.longestValidParentheses(s);
    cout << ret << endl;
    return 0;
}
